_posts/2021-01-13-hapi-fhir.md

```diff
@@ -0,0 +1,454 @@
+HAPI uses annotations to indicate what service the procedures provide. 
+The procedure createPatient in the diagram is annotated with @Create which indicates it handles POST/create. 
+This procedure then uses a PatientDAO (DAO - data access object) class which uses Spring Data to persist the Patient resource in the MongoDB.
+
+  
+
+```Java
+import ca.uhn.fhir.context.FhirContext;
+import ca.uhn.fhir.rest.annotation.*;
+import ca.uhn.fhir.rest.api.MethodOutcome;
+import ca.uhn.fhir.rest.param.DateRangeParam;
+import ca.uhn.fhir.rest.param.StringParam;
+import ca.uhn.fhir.rest.param.TokenParam;
+import ca.uhn.fhir.rest.server.IResourceProvider;
+import com.example.fhirexample.dao.patient.IPatient;
+import org.hl7.fhir.instance.model.api.IBaseResource;
+import org.hl7.fhir.r4.model.IdType;
+import org.hl7.fhir.r4.model.OperationOutcome;
+import org.hl7.fhir.r4.model.Patient;
+import org.hl7.fhir.r4.model.Resource;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import javax.servlet.http.HttpServletRequest;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+@Component
+public class PatientProvidor implements IResourceProvider {
+
+    @Autowired
+    private FhirContext ctx;
+
+    @Autowired
+    private IPatient patientDao;
+
+    private Map<String, Patient> patientMap = new HashMap<String, Patient>();
+
+    public  PatientProvidor() {
+    
+    }   
+    
+    @Override
+    public Class<? extends IBaseResource> getResourceType() {
+        return  Patient.class;
+    }
+
+    @Search
+    public List<Resource> searchPatient(HttpServletRequest request,
+                                        @OptionalParam(name= Patient.SP_BIRTHDATE) DateRangeParam birthDate,
+                                        @OptionalParam(name = Patient.SP_FAMILY) StringParam familyName,
+                                        @OptionalParam(name= Patient.SP_GENDER) StringParam gender,   
+                                        @OptionalParam(name= Patient.SP_GIVEN) StringParam givenName,
+                                        @OptionalParam(name = Patient.SP_IDENTIFIER) TokenParam identifier,
+                                        @OptionalParam(name= Patient.SP_NAME) StringParam name,
+                                        @OptionalParam(name = Patient.SP_RES_ID) TokenParam resid) {
+
+        List<Resource> results = patientDao.search(ctx, birthDate, familyName, gender, givenName, identifier, name);
+        return results;
+    }
+
+    @Read()
+    public Patient read(@IdParam IdType theId) {
+
+        Patient patient = patientDao.read(ctx,theId);
+        return patient;
+    }
+
+    @Create()
+    public MethodOutcome createPatient(HttpServletRequest theRequest, @ResourceParam Patient patient) {
+
+        MethodOutcome  method = new  MethodOutcome();
+        method.setCreated(true);
+        OperationOutcome  opOutcome = new  OperationOutcome();
```

Comment:
- What's the use of OperationOutcome? No need to call setters like for the MethodOutcome?
- Why does only this method has a MethodOutcome/OperationOutcome? Shouldn't the other methods in this class also return this?

Reply:
The MethodOutcome object must be returned on update and create methods. This object contains the identity of the created resource.  On a delete and validate method  you have a choice between void and MethodOutcome. This is mentioned in the FHIR documentation (https://hapifhir.io/hapi-fhir/docs/server_plain/rest_operations.html).
Operation outcomes are sets of error, warning and information messages that provide detailed information about the outcome of an attempted system operation. They are provided as a direct system response, or component of one, and provide information about the outcome of the operation